dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,640 @@
+
+            signature = inspect.Signature(parameters)
+
+            def call_impl(self, *args, **kwargs):
+                try:
+                    bound_args = signature.bind(self, *args, **kwargs)
+                except Exception as exc:
+                    raise RuntimeError(f"{exc}. Signature is {signature}")
+                return self._add_sample(bound_args)
+
+            call_impl.__signature__ = inspect.Signature(parameters)
+            _DALIProxy.__call__ = call_impl
+            self._proxy = _DALIProxy(
+                self._dali_input_names,
+                self._dali_input_q,
+                self._pipe.num_outputs,
+                self._deterministic,
+            )
+
+        return self._proxy
+
+    def _schedule_batch(self, pipe_run_ref):
+        _torch.cuda.nvtx.range_push(f"dali_proxy.dali_input_q.put {pipe_run_ref.batch_id}")
+        self._dali_input_q.put((pipe_run_ref.batch_id, pipe_run_ref.inputs))
+        pipe_run_ref.is_scheduled = True
+        _torch.cuda.nvtx.range_pop()
+
+    def _get_outputs(self, pipe_run_ref: DALIPipelineRunRef):
+        """
+        Gets the pipeline outputs for a specific batch id. It will keep reading data until the
+        right batch is found, caching the results that were not consumed until a future call
+        """
+        req_batch_id = pipe_run_ref.batch_id
+
+        # In case we haven't scheduled the iteration yet (i.e. deterministic config), do it now
+        if not pipe_run_ref.is_scheduled:
+            self._schedule_batch(pipe_run_ref)
+
+        # Wait for the requested output to be ready
+        req_outputs = None
+        # If the data was already read, just return it (and clear the cache entry)
+        if req_batch_id in self._cache_outputs:
+            req_outputs = self._cache_outputs[req_batch_id]
+            del self._cache_outputs[req_batch_id]
+
+        else:
+            curr_batch_id = None
+            # If not the data we are looking for, store it and keep processing until we find it
+            while req_batch_id != curr_batch_id:
+                _torch.cuda.nvtx.range_push("dali_output_q.get")
+                curr_batch_id, curr_processed_outputs, err = self._dali_output_q.get()
+                _torch.cuda.nvtx.range_pop()
+
+                if err is not None:
+                    raise err
+
+                if curr_batch_id == req_batch_id:
+                    req_outputs = curr_processed_outputs
+                else:
+                    self._cache_outputs[curr_batch_id] = curr_processed_outputs
+        return req_outputs
+
+    def _produce_data_impl(self, obj, cache):
+        """
+        Recursive implementation of produce_data
+        """
+        # If it is an instance of DALIOutputBatchRef, replace it with data
+        if isinstance(obj, DALIOutputBatchRef):
+            if obj.pipe_run_ref.batch_id not in cache:
+                cache[obj.pipe_run_ref.batch_id] = self._get_outputs(obj.pipe_run_ref)
```

Comment:
Just to be clear, can we expect here results from more than one batch / run? Wouldn't that mean that something went wrong? Or is it ok for the user to keep several pipe_run_refs, group them and process at once?

Reply:
should not happen because we query the pipeline for all the outputs in the pipeline and either consume themm all or store them in the cache